xformers/ops/fmha/tensor_with_seqlen.py

```diff
@@ -108,31 +119,41 @@ def forward(
                     max_seqlen=max_seqlen,
                     cu_seqlen=cu_seqlen,
                     cu_seqlen_py=cu_seqlen_py,
+                    batch_sizes=batch_sizes,
                 )
 
             @staticmethod
             @torch.autograd.function.once_differentiable
             def backward(ctx, grad: torch.Tensor):
-                return grad, None, None
+                return grad, None, None, None
 
-        c = torch.cat(tuple(tensors), dim=dim)
+        tensors_bs1 = tuple(x.reshape([1, -1, *x.shape[2:]]) for x in tensors)
+        c = torch.cat(tensors_bs1, dim=1)
         cu_seqlen = [0]
+        batch_sizes = []
         max_seqlen = -1
         for tensor in tensors:
             assert not isinstance(tensor, TensorWithSeqLen)
-            seqlen = tensor.shape[dim]
+            seqlen = tensor.shape[1]
             max_seqlen = max(max_seqlen, seqlen)
-            cu_seqlen.append(cu_seqlen[len(cu_seqlen) - 1] + seqlen)
-        return FromTensorListOp.apply(c, max_seqlen, cu_seqlen)
+            for _ in range(tensor.shape[0]):
+                cu_seqlen.append(cu_seqlen[len(cu_seqlen) - 1] + seqlen)
+            batch_sizes.append(tensor.shape[0])
+        return FromTensorListOp.apply(c, max_seqlen, cu_seqlen, batch_sizes)
 
-    def to_tensor_list(self, dim: int) -> List[torch.Tensor]:
-        if self.cu_seqlen_py[-1] != self.shape[dim]:
+    def to_tensor_list(self) -> List[torch.Tensor]:
+        if self.cu_seqlen_py[-1] != self.shape[1] or self.shape[0] != 1:
             raise ValueError(
                 f"Invalid `TensorWithSeqLen` of shape {self.shape}.\n"
                 f" cu_seqlen: {self.cu_seqlen_py}"
             )
         out = []
-        for begin, end in zip(self.cu_seqlen_py, self.cu_seqlen_py[1:]):
-            s = self[(slice(None),) * dim + (slice(begin, end),)]
-            out.append(s)
+        it = 0
+        for batch_size in self.batch_sizes:
+            this_tensor = []
+            for _ in range(batch_size):
+                begin, end = self.cu_seqlen_py[it], self.cu_seqlen_py[it + 1]
+                this_tensor.append(self[:, begin:end])
+                it += 1
```

Comment:
nit: this can be implemented with `tensor.split`, by passing the `batch_sizes * seq_len` as its arguments, followed by a reshape. It will be probably a bit faster as well I believe